src/storage/in_mem_csv_copier/in_mem_arrow_node_copier.cpp

```diff
@@ -0,0 +1,626 @@
+                    InMemColumnFactory::getInMemPropertyColumn(fName, property.dataType, numNodes);
+        }
+        logger->info("Done initializing in memory structured columns.");
+    }
+
+
+    template<typename T>
+    arrow::Status InMemArrowNodeCopier::populateColumns() {
+        logger->info("Populating structured properties");
+        auto pkIndex =
+                make_unique<HashIndexBuilder<T>>(StorageUtils::getNodeIndexFName(this->outputDirectory,
+                                                                                 nodeTableSchema->tableID,
+                                                                                 DBFileType::WAL_VERSION),
+                                                 nodeTableSchema->getPrimaryKey().dataType);
+        pkIndex->bulkReserve(numNodes);
+
+        arrow::Status status;
+        switch (inputFileType) {
+            case FileTypes::CSV:
+                status = populateColumnsFromCSV<T>(pkIndex);
+                break;
+
+            case FileTypes::ARROW:
+                status = populateColumnsFromArrow<T>(pkIndex);
+                break;
+
+            case FileTypes::PARQUET:
+                status = populateColumnsFromParquet<T>(pkIndex);
+                break;
+        }
+
+        logger->info("Flush the pk index to disk.");
+        pkIndex->flush();
+        logger->info("Done populating structured properties, constructing the pk index.");
+
+        return status;
+    }
+
+    template<typename T>
+    arrow::Status InMemArrowNodeCopier::populateColumnsFromCSV(unique_ptr<HashIndexBuilder<T>> &pkIndex) {
+        node_offset_t offsetStart = 0;
+
+        shared_ptr<arrow::csv::StreamingReader> csv_streaming_reader;
+        auto status = initCSVReader(csv_streaming_reader, csvDescription.filePath);
+        if (!status.ok()) {
+            return status;
+        }
+
+        std::shared_ptr<arrow::RecordBatch> currBatch;
+
+        int blockIdx = 0;
+        auto it = csv_streaming_reader->begin();
+        auto endIt = csv_streaming_reader->end();
+        while (it != endIt) {
+            for (int batchOfTasksId = 0; batchOfTasksId < CopyCSVConfig::NUM_TASKS_PER_BATCH; ++ batchOfTasksId) {
+                if (it == endIt) {
+                    break;
+                }
+                ARROW_ASSIGN_OR_RAISE(currBatch, *it);
+                taskScheduler.scheduleTask(CopyCSVTaskFactory::createCopyCSVTask(
+                        batchPopulateColumnsTask<T, arrow::Array>,
+                        nodeTableSchema->primaryKeyPropertyIdx,
+                        blockIdx, offsetStart, pkIndex.get(), this, currBatch->columns(),
+                        csvDescription.csvReaderConfig.tokenSeparator));
+                offsetStart += currBatch->num_rows();
+                ++blockIdx;
+                ++ it;
+            }
+            while (taskScheduler.getTaskNum() > CopyCSVConfig::MINIMUM_TASKS_TO_SCHEDULE_MORE) {
```

Comment:
Let's refactor this to waitUntilEnoughTasksFinish() across the 3 implementations.